Converters/Tekla/Speckle.Converters.TeklaShared/ToSpeckle/Raw/OpeningToSpeckleConverter.cs

```diff
@@ -0,0 +1,157 @@
+using Speckle.Converters.Common;
+using Speckle.Converters.Common.Objects;
+using Speckle.Sdk.Models;
+
+namespace Speckle.Converters.TeklaShared.ToSpeckle.Raw;
+
+[NameAndRankValue(nameof(TSM.BooleanPart), NameAndRankValueAttribute.SPECKLE_DEFAULT_RANK)]
+public class OpeningToSpeckleConverter : ITypedConverter<TSM.BooleanPart, IEnumerable<Base>>
+{
+  private readonly IConverterSettingsStore<TeklaConversionSettings> _settingsStore;
+  private readonly ITypedConverter<TG.LineSegment, SOG.Line> _lineConverter;
+
+  public OpeningToSpeckleConverter(
+    IConverterSettingsStore<TeklaConversionSettings> settingsStore,
+    ITypedConverter<TG.LineSegment, SOG.Line> lineConverter
+  )
+  {
+    _settingsStore = settingsStore;
+    _lineConverter = lineConverter;
+  }
+
+  private (double minZ, double maxZ) GetParentZBounds(TSM.ModelObject parent)
+  {
+    TSM.Solid? solid = null;
+
+    // detect if the solid is a part or cut
+    if (parent is TSM.Part part)
+    {
+      solid = part.GetSolid();
+    }
+    else if (parent is TSM.BooleanPart boolPart)
+    {
+      solid = boolPart.OperativePart?.GetSolid();
+    }
+
+    if (solid == null)
+    {
+      return (0, 0);
+    }
+
+    double minZ = double.MaxValue;
+    double maxZ = double.MinValue;
+
+    var faceEnum = solid.GetFaceEnumerator();
+    while (faceEnum.MoveNext())
+    {
+      var face = faceEnum.Current;
+      if (face == null)
+      {
+        continue;
+      }
+
+      var loopEnum = face.GetLoopEnumerator();
+      while (loopEnum.MoveNext())
+      {
+        var loop = loopEnum.Current;
+        if (loop == null)
+        {
+          continue;
+        }
+
+        var vertexEnum = loop.GetVertexEnumerator();
+        while (vertexEnum.MoveNext())
```

Comment:
OMG. 3 nested while loop, how crazy we are ( ͡°( ͡° ͜ʖ( ͡° ͜ʖ ͡°)ʖ ͡°) ͡°)
at least this converter deserves inline comments/notes. hard to understand from first look. also wanna know why we need that while loops, seems like API thing but some explanation around structure would be good